python/tank/bootstrap/manager.py

```diff
@@ -221,49 +247,160 @@ def set_progress_callback(self, callback):
         """
         self._progress_cb = callback
 
-    def bootstrap_engine(self, engine_name, entity=None):
+    def bootstrap_engine(self,
+                         engine_name,
+                         entity=None,
+                         progress_callback=None,
+                         completed_callback=None,
+                         failed_callback=None):
         """
         Create an sgtk instance for the given project or site,
         then launch into the given engine.
 
+        When property ``async_bootstrap`` is false, the whole engine bootstrap logic
+        will be executed synchronously in the main application thread. When true,
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
         If entity is None, the method will bootstrap into the site
         config. This method will attempt to resolve the config according
         to business logic set in the associated resolver class and based
         on this launch a configuration. This may involve downloading new
         apps from the toolkit app store and installing files on disk.
 
+        3 callback functions can be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(step_number, message, current_index, maximum_index)
```

Comment:
wondering if we need step number here - this would mean making breaking changes to `set_progress_callback`. The idea with the existing progress callback signature `progress_callback(message, current_index, max_index)` is that it's hard to give a global overview of "how much time is left until we are done". Instead calls are issued along the lines of:

```
progress_callback("Setting up...", 1, 1)
progress_callback("Swapping core...", 1, 1)
progress_callback("Downloading x...", 1, 100)
progress_callback("Downloading y...", 2, 100)
progress_callback("Downloading z...", 3, 100)
...
progress_callback("Downloading w...", 100, 100)
progress_callback("Starting engine", 1, 1)
```

Does that still make sense you think?